### PR TITLE
Added retry logic for headless fetches

### PR DESCRIPTION
**Pull request by chunder**
I noticed that WA, USA was flakey. The server or headless seems to fail <10% of the time. So I added retry logic. I couldn't get any failures to happen to prove it out though. So can someone review this code carefully?

### CODE REVIEW

File: src/events/crawler/lib/fetch.js [hunk 1]
```diff
@@ -147,30 +147,49 @@ const fetchHeadless = async url => {
...
+        await page.waitFor(RESPONSE_TIMEOUT);
+        const html = await page.content();
+        browser.close();
+        return html;
+      }
+      if (response._status < 500) {
+        console.log(`  ❌ Got error ${response._status} trying to fetch ${url}`);
+        browser.close();
+        return null;
+      }
```
Comment (jzohrab) [inline src/events/crawler/lib/fetch.js]: Will the server ever have an internal server error (code >= 500)?  And if so, should it retry?  E.g., see the checks in https://github.com/lazd/coronadatascraper/blob/master/src/events/crawler/lib/get.js#L80.

File: src/events/crawler/lib/fetch.js [hunk 2]
```diff
@@ -147,30 +147,49 @@ const fetchHeadless = async url => {
...
-  } catch (err) {
-    browser.close();
 
-    if (err.name === 'TimeoutError') {
-      console.log(`  ❌ Timed out trying to fetch ${url}`);
-      return null;
+      if (err.name === 'TimeoutError') {
+        console.log(`  ❌ Timed out trying to fetch ${url}`);
+      } else {
+        console.log(`  ❌ Caught error trying to fetch ${url}`);
```
Comment (jzohrab) [inline src/events/crawler/lib/fetch.js]: Consider actually writing error details here.

File: src/events/crawler/lib/fetch.js [hunk 3]
```diff
@@ -147,30 +147,49 @@ const fetchHeadless = async url => {
...
+      }
     }
-    throw err;
   }
+
+  return null;
```
Comment (jzohrab) [inline src/events/crawler/lib/fetch.js]: Consider logging something to the effect of "exceeded max retries" here.

**Comment by jzohrab** [comment]
Hi @chunder, I don't know that tech well, so my notes might not be the "careful code review" you're looking for.  Cheers! z

File: src/events/crawler/lib/fetch.js (hunk 1 above)
Comment (jzohrab) [reply src/events/crawler/lib/fetch.js]: (I'm assuming that a >=500 error wouldn't actually result in an error being thrown)

**Comment by chunder** [comment]
@jzohrab I added better comments and logging per your suggestions.

@lazd @qgolsteyn Either you able to eyeball the logic here. I think it's sane, but it's hard to test.

File: src/events/crawler/lib/fetch.js (hunk 1 above)
Comment (lazd) [reply src/events/crawler/lib/fetch.js]: Hmm, I think retrying on an internal server error is likely to just result in another failure (trying 30 minutes later, maybe not), but it's possible we could get it right as it's fixed...

File: src/events/crawler/lib/fetch.js [hunk 4]
```diff
@@ -147,30 +147,50 @@ const fetchHeadless = async url => {
...
-      waitUntil: 'networkidle2'
-    });
-
-    if (response._status < 400) {
-      await page.waitFor(RESPONSE_TIMEOUT);
-      const html = await page.content();
-      browser.close();
-      return html;
+  let tries = 0;
+  while (tries < 5) {
```
Comment (lazd) [inline src/events/crawler/lib/fetch.js]:
```suggestion
  let tries = 0;
  let maxTries = 5;
  while (tries < maxTries) {
```

File: src/events/crawler/lib/fetch.js (hunk 1 above)
Comment (chunder) [reply src/events/crawler/lib/fetch.js]: Servers return 503 or other 5xx all the time when they are simply overloaded. Sleeping a bit and retrying is a perfectly great thing to do.

File: src/events/crawler/lib/fetch.js [hunk 5]
```diff
@@ -147,30 +147,50 @@ const fetchHeadless = async url => {
...
+      }
+    } catch (err) {
+      // Caught something, allow retry
+      browser.close();
+      console.log(`  ❌ Caught ${err.name} error trying to fetch ${url}`);
     }
-    throw err;
   }
+
+  console.log(`  ❌ Too many trying to fetch ${url}`);
```
Comment (lazd) [inline src/events/crawler/lib/fetch.js]:
```suggestion
  console.log(`  ❌ Failed to fetch ${url} after ${tries} tries`);
```

**Comment by lazd** [comment]
@chunder it does indeed look sane. I made this and the other retry logic the same and added better logging, can you look again at it and check I didn't screw anything up?

**Pull request merged**